src/null-safety/understanding-null-safety/index.md

```diff
@@ -0,0 +1,1410 @@
+classes like your own types near the bottom.
+
+If that directed graph comes to a point at the top where there is a single type
+that is the supertype (directly or indirectly), that type is called the *top
+type*. Likewise, if there is a weird type at that bottom that is a subtype of
+every type, you have a *bottom type*. (In this case, your directed graph is a
+[lattice][].)
+
+[lattice]: https://en.wikipedia.org/wiki/Lattice_(order)
+
+It's convenient if your type system has a top and bottom type, because it means
+that type-level operations like least upper bound (which type inference uses to
+figure out the type of a conditional expression based on the types of its two
+branches) can always produce a type. Before null safety, `Object` was Dart's top
+type and `Null` was its bottom type.
+
+Since `Object` is non-nullable now, it is no longer a top type. `Null` is not a
+subtype of it. Dart has no *named* top type. If you need a top type, you want
+`Object?`. Likewise, `Null` is no longer the bottom type. If it was, everything
+would still be nullable. Instead, we've added a new bottom type named `Never`:
+
+<img src="understanding-null-safety/top-and-bottom.png" width="50%">
+
+In practice, this means:
+
+*   If you want to indicate that you allow a value of any type, use `Object?`
+    instead of `Object`. In fact, it becomes pretty rare to use `Object` since
+    that type means "could be any possible value except this one weirdly
+    prohibted value `null`".
+
+*   On the rare times you need a bottom type, use `Never` instead of `Null`.
+    If you don't know if you need it, you probably don't.
+
+## Ensuring correctness
+
+We bifurcated the universe of types into nullable and non-nullable halves. In
+order to maintain soundness and our principle that you can never get a null
+reference error at runtime unless you ask for it, we need to guarantee that
+`null` never appears in any type on the non-nullable side.
+
+Getting rid of implicit downcasts and removing `Null` as a bottom type covers
+all of the main places that types flow through a program across assigments and
+from arguments into parameters on function calls. The main remaining places
+where `null` can sneak in are when a variable first comes into being and when
+you leave a function. So there are some additional compile errors:
+
+### Invalid returns
+
+If a function has a non-nullable return type, then every path through the
+function must reach a `return` statement that returns a value. Before null
+safety, Dart was pretty lax about missing returns. For example:
+
+```dart
+String missingReturn() {
+  // No return.
+}
+```
+
+If you analyze this, you get a gentle *hint* that *maybe* you forgot a return,
+but if not, no big deal. That's because if execution reaches the end of a
+function body then Dart implicitly returns `null`. Since every type is nullable,
+*technically* this function is safe, even though it's probably not what you
+want.
```

Comment:
I wonder if it makes sense to use past tense in paragraphs like this. People scan documents, or they just don't get the meaning of a sentence above, and they might get the impression that the behavior explained here is the current one. It might help to put all the "before null safety" in past tense, and all the "after" as present tense.

This will also make the document more long-lived as-is. I think we expect most readers to read this when null safety is either almost here, or completely here.

Reply:
Yeah, tense and temporality is hard. I'm trying to balance both where the language is at the point in time that a reader reads this, and where they are as they make progress through the document. Also, large passages in convoluted tenses can be painful to read. ("Large passages that had been being in convoluted tenses will have been painful to have been reading.")

You're right this paragraph reads better as past tense. Done. Tweaked some other places too.